proxyfinder.go

```diff
@@ -85,34 +89,36 @@ func (pf *ProxyFinder) findProxyForRequest(req *http.Request) (*url.URL, error)
 		log.Printf(`[%d] %s %s via "DIRECT" (not connected)`, id, req.Method, req.URL)
 		return nil, nil
 	}
-	s, err := pf.runner.FindProxyForURL(req.URL)
+	str, err := pf.runner.FindProxyForURL(req.URL)
 	if err != nil {
 		return nil, err
 	}
-	log.Printf("[%d] %s %s via %q", id, req.Method, req.URL, s)
-	ss := strings.Split(s, ";")
-	if len(ss) > 1 {
-		log.Printf("[%d] Warning: ignoring all but first proxy in %q", id, s)
-	}
-	trimmed := strings.TrimSpace(ss[0])
-	if trimmed == "DIRECT" {
-		return nil, nil
-	}
-	var host string
-	n, err := fmt.Sscanf(trimmed, "PROXY %s", &host)
-	if err == nil && n == 1 {
-		// The specified proxy should contain both a host and a port, but if for some reason
-		// it doesn't, assume port 80. This needs to be made explicit, as it eventually gets
-		// passed to net.Dial, which also requires a port.
-		proxy := &url.URL{Host: host}
-		if proxy.Port() == "" {
-			proxy.Host = net.JoinHostPort(host, "80")
+	arr := strings.Split(str, ";")
+	for _, elem := range arr {
+		fields := strings.Fields(strings.TrimSpace(elem))
+		if len(fields) == 1 && fields[0] == "DIRECT" {
+			log.Printf("[%d] %s %s via %q", id, req.Method, req.URL, elem)
+			return nil, nil
+		} else if len(fields) == 2 && fields[0] == "PROXY" {
+			// The specified proxy should contain both a host and a port, but if for
+			// some reason it doesn't, assume port 80. This needs to be made explicit,
+			// as it eventually gets passed to net.Dial, which also requires a port.
+			proxy := &url.URL{Host: fields[1]}
+			if proxy.Port() == "" {
+				proxy.Host = net.JoinHostPort(proxy.Host, "80")
+			}
+			if pf.blocked.contains(proxy.Host) {
+				log.Printf("[%d] Skipping bad proxy: %q", id, proxy.Host)
```

Comment:
this will produce a bit og log spam wont it? it will log on every request through alpaca when the first proxy on the list goes bad, until it comes good again?

I think instead you should just log when you put an entry on the block list and when you expire them.

Reply:
That's a good point. I've logged this in both places before we block a proxy (rather than in `proxyFinder.blockProxy()`, because I also want to log the id).